Add stream search from clickhouse tutorial (#897)
quickwit-inc/quickwit: docs/tutorials/add-full-text-search-to-your-olab-db.md

> ---

Reviewer: the file name `add-full-text-search-to-your-olab-db` -> olap instead of olab.

title: Add full-text search to Clickhouse
sidebar_position: 2
---

> This guide will help you add full-text search to a well-known OLAP database, Clickhouse, using the Quickwit search streaming feature. Indeed Quickwit exposes a REST endpoint that streams ids or whatever attributes matching a search query **extremely fast** (up to 50 million in 1 second), and Clickhouse can easily use them with joins queries. 

Reviewer: ```suggestion
This guide will help you add full-text search to a well-known OLAP database, Clickhouse, using the Quickwit search streaming feature. Indeed Quickwit exposes a REST endpoint that streams ids or whatever attributes matching a search query **extremely fast** (up to 50 million/second), and Clickhouse can easily use them with `JOIN` queries. 
```

> We will take the [Github archive dataset](https://www.gharchive.org/), which gathers more than 3 billion Github events: `WathEvent`, `PullRequestEvent`, `IssuesEvent`... You can dive into this [great analysis] (https://ghe.clickhouse.tech/) made by Clickhouse to have a good understanding of the dataset. We also took strong inspiration from this work, and we are very grateful for them to have shared this.

Reviewer: ```suggestion
We will take the [Github archive dataset](https://www.gharchive.org/), which gathers more than 3 billion Github events: `WathEvent`, `PullRequestEvent`, `IssuesEvent`... You can dive into this [great analysis](https://ghe.clickhouse.tech/) made by Clickhouse to have a good understanding of the dataset. We also took strong inspiration from this work, and we are very grateful for them to have shared this.
```

## Install

```bash
curl -L https://install.quickwit.io | sh

> cd quickwit-0.2

Reviewer: ```suggestion
cd quickwit-v0.2.0
```

Reviewer: We can also do `cd quickwit-v*/`. Whatever you decide, please let me know so we can be consistent in other parts of the docs.

Author: let's go for `cd quickwit-v*/`

  "ref": null,
  "additions": 5,
  "deletions": 5,
  "commit_id": null,
  "body":"This PR contains the following updates..."
}
```

We don't need to index all fields described above as Quickwit. The `title` and `body` are the fields of interest for our full-text search tutorial. `id` will be helpful to make the join in Clickhouse, `created_at` and `event_type` may also be beneficial for timestamp pruning and filtering.

```yaml title="gh-archive-index-config.yaml"
version: 0
index_id: gh-archive
index_uri: INDEX_URI
doc_mapping:
  field_mappings:
    - name: id
      type: u64
      fast: true
      stored: true
    - name: created_at
      type: i64
      fast: true
      stored: true
    - name: event_type
      type: text
      tokenizer: raw
      stored: true
    - name: title
      type: text
      tokenizer: default
      record: position
      stored: true
    - name: body
      type: text
      tokenizer: default
      record: position
      stored: true
search_settings:
  default_search_fields: [title, body]
}
```

```bash
curl -o gh-archive-index-config.yaml https://raw.githubusercontent.com/quickwit-inc/quickwit/main/config/tutorials/gh-archive/index-config.yaml

> quickwit index create --index gh-archive --index-config gh-archive-index-config.yaml

Reviewer: ```suggestion
./quickwit index create --index gh-archive --index-config gh-archive-index-config.yaml
```

```

## Indexing events

The dataset is a compressed [ndjson file](https://quickwit-datasets-public.s3.amazonaws.com/gh-archive/gh-archive-2021-12.json.gz).
Let's index it.

```bash
curl https://quickwit-datasets-public.s3.amazonaws.com/gh-archive/gh-archive-2021-12-text-only.json.gz

> gunzip gh-archive-2021-12-text-only.json.gz | quickwit index ingest --index gh-archive

Reviewer: ```suggestion
gunzip gh-archive-2021-12-text-only.json.gz | ./quickwit index ingest --index gh-archive
```

```

You can check it's working by using the `search` command and looking for `tantivy` word:
```bash

> quickwit index search --index gh-archive --query "tantivy"

Reviewer: ```suggestion
./quickwit index search --index gh-archive --query "tantivy"
```

```

## Start a searcher

```bash

> quickwit service run searcher

Reviewer: ```suggestion
./quickwit service run searcher
```

    actor_login LowCardinality(String),
    repo_name LowCardinality(String),
    created_at Int64,
    action Enum('none' = 0, 'created' = 1, 'added' = 2, 'edited' = 3, 'deleted' = 4, 'opened' = 5, 'closed' = 6, 'reopened' = 7, 'assigned' = 8, 'unassigned' = 9,
                'labeled' = 10, 'unlabeled' = 11, 'review_requested' = 12, 'review_request_removed' = 13, 'synchronize' = 14, 'started' = 15, 'published' = 16, 'update' = 17, 'create' = 18, 'fork' = 19, 'merged' = 20),
    comment_id UInt64,
    body String,
    ref LowCardinality(String),
    number UInt32,
    title String,
    labels Array(LowCardinality(String)),
    additions UInt32,
    deletions UInt32,
    commit_id String,
) ENGINE = MergeTree ORDER BY (type, repo_name, created_at);
```

### Import events

We have created a second dataset, `gh-archive-2021-12.json.gz`, which gathers all events, even ones with no
text. So it's better to insert it into Clickhouse, but if you don't have the time, you can use the dataset
`gh-archive-2021-12-text-only.json.gz` used for Quickwit.

```bash
curl https://quickwit-datasets-public.s3.amazonaws.com/gh-archive/gh-archive-2021-12.json.gz

> gunzip . | clickhouse-client -d gh-archive --query="INSERT INTO github_events FORMAT JSONEachRow"

Reviewer: ```suggestion
gunzip gh-archive-2021-12.json.gz | clickhouse-client -d gh-archive --query="INSERT INTO github_events FORMAT JSONEachRow"
```

```

Let's check it's working:
```SQL

> // Top repositories by stars

Reviewer: ```suggestion
-- Top repositories by stars
```

│ Eugeny/tabby         │  4027 │
│ prabhatsharma/zinc   │  3936 │
│ AppFlowy-IO/appflowy │  3382 │
│ vercel/turborepo     │  3314 │
└──────────────────────┴───────┘
```

### Use Quickwit search inside Clickhouse

Clikhouse has an exciting feature called [URL Table Engine](https://clickhouse.com/docs/en/engines/table-engines/special/url/) that queries data from a remote HTTP/HTTPS server.
This is precisely what we need: by creating a table pointing to Quickwit search stream endpoint, we will fetch ids that match a query from Clickhouse. 

```SQL
SELECT count(*) FROM url('http://127.0.0.1:7280/api/v1/gh-archive/search/stream?query=log4j+OR+log4shell&fastField=id&outputFormat=clickHouseRowBinary', RowBinary, 'id UInt64')

┌─count()─┐
│   99584 │
└─────────┘

1 rows in set. Elapsed: 0.012 sec. Processed 99.13 thousand rows, 793.00 KB (7.96 million rows/s., 63.66 MB/s.)
```

We are fetching 100 000 u64 ids in 0.012 seconds. That's 8 million rows per second, not bad. And it's possible to increase the throughput with more extensive queries.

Let's do another example with a more exciting query that will match `log4j` or `log4shell` and count events per day:

> SELECT

Reviewer: ```suggestion
```SQL
SELECT
```
